lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/FileBasedSystemAccessControl.java

```diff
@@ -963,18 +988,30 @@ public void checkCanExecuteProcedure(SystemSecurityContext systemSecurityContext
     @Override
     public void checkCanExecuteFunction(SystemSecurityContext systemSecurityContext, String functionName)
     {
+        Identity identity = systemSecurityContext.getIdentity();
+        boolean allowed = functionRules.stream()
+                .filter(functionRule -> functionRule.matches(identity.getUser(), identity.getEnabledRoles(), identity.getGroups(), functionName))
+                .map(CatalogFunctionAccessControlRule::canExecuteFunction)
+                .findFirst()
+                .orElse(false);
+        if (!allowed) {
+            denyExecuteFunction(functionName);
+        }
```

Comment:
I'd try to extract the whole body of this method to another one, which would be reused in the method below.
To make that possible the additional method should be provided to redirect to `functionRule.matches` with/without  functionKind depending if optional isPresent 

that would underline that the logic is the same. in both methods.
However I'm not sure if that would be more clumsy or not.

Reply:
Extracted, I think it looks better